python/ray/autoscaler/azure/config.py

```diff
@@ -59,169 +57,60 @@ def _configure_resource_group(config):
-            assignments = auth_client.role_assignments.list_for_scope(
-                scope=resource_group_id, filter=filter_expr)
-
-            if any(a.role_definition_id == role_id for a in assignments):
-                break
-
-            auth_client.role_assignments.create(
-                scope=resource_group_id,
-                role_assignment_name=uuid.uuid4(),
-                parameters=role_params)
-            logger.info("Assigning Contributor Role to MSI User")
-        except CloudError as ce:
-            if ce.inner_exception.error == "PrincipalNotFound":
-                time.sleep(5)
-    else:
-        raise Exception(
-            "Failed to create contributor role assignment (timeout)")
+    deployment_async_operation = resource_client.deployments.create_or_update(
+        resource_group, 'ray-config', deployment_properties)
+    deployment_async_operation.wait()
 
     return config
 
 
 def _configure_key_pair(config):
     ssh_user = config["auth"]["ssh_user"]
+    private_key_path = config["auth"].get("ssh_private_key")
+    if private_key_path:
+        # skip key generation if it is manually specified
+        assert os.path.exists(private_key_path), (
+            "Could not find private ssh key: {}".format(private_key_path))
+
+        # make sure public key also exists
+        public_key_path = config["auth"]["ssh_public_key"]
+        assert os.path.exists(public_key_path), (
+            "Could not find public ssh key: {}".format(public_key_path))
+    else:
+        resource_group = config["provider"]["resource_group"]
 
-    for key_type in ["ssh_private_key", "ssh_public_key"]:
-        try:
-            key_path = os.path.expanduser(config["auth"][key_type])
-        except KeyError:
-            raise Exception("Config must define {}".format(key_type))
-        except TypeError:
-            raise Exception("Invalid config value for {}".format(key_type))
-
-        assert os.path.exists(key_path), (
-            "Could not find ssh key: {}".format(key_path))
-
-        if key_type == "ssh_public_key":
-            with open(key_path, "r") as f:
-                public_key = f.read()
-
-    os_profile = {
-        "admin_username": ssh_user,
-        "computer_name": None,
-        "linux_configuration": {
-            "disable_password_authentication": True,
-            "ssh": {
-                "public_keys": [{
-                    "key_data": public_key,
-                    "path": "/home/{}/.ssh/authorized_keys".format(ssh_user)
-                }]
-            }
-        }
-    }
-    for node_type in ["head_node", "worker_nodes"]:
-        config[node_type]["os_profile"] = os_profile
-
-    return config
+        # look for an existing key pair
+        key_name = "ray_azure_{}_{}".format(resource_group, ssh_user)
+        public_key_path = os.path.expanduser("~/.ssh/{}.pub".format(key_name))
+        private_key_path = os.path.expanduser("~/.ssh/{}.pem".format(key_name))
```

Comment:
I updated this section to enforce manual creation of ssh key-pair prior to setup.
The expectation now is that the config auth section will provide entries for ssh_private_key and ssh_public_key which will have values that are the paths to those keys